cockatrice/src/pictureloader.cpp

```diff
@@ -43,22 +43,96 @@ class PictureToLoad::SetDownloadPriorityComparator
     }
 };
 
-PictureToLoad::PictureToLoad(CardInfoPtr _card) : card(std::move(_card)), setIndex(0)
+PictureToLoad::PictureToLoad(CardInfoPtr _card) :
+    card(std::move(_card)),
+    setIndex(0),
+    urlIndex(0)
 {
+    // This will be replaced with an expandable list ideally
+    urlTemplates.append(settingsCache->getPicUrl());
+    urlTemplates.append(settingsCache->getPicUrlFallback());
+
     if (card) {
         sortedSets = card->getSets();
         qSort(sortedSets.begin(), sortedSets.end(), SetDownloadPriorityComparator());
+        populateSetUrls();
+    }
+}
+
+void PictureToLoad::populateSetUrls()
+{
+    currentSetUrls.clear();
+
+    QString setCustomURL = card->getCustomPicURL(getCurrentSet()->getShortName());
+
+    if (!setCustomURL.isEmpty()) {
+        currentSetUrls.append(setCustomURL);
+    }
+
+    for (int i; i< urlTemplates.size(); i++) {
+        QString transformedUrl = transformUrl(urlTemplates[i]);
+
+        if (!transformedUrl.isEmpty()) {
+            currentSetUrls.append(transformedUrl);
+        }
     }
+
+    urlIndex = 0;
 }
 
 bool PictureToLoad::nextSet()
 {
     if (setIndex == sortedSets.size() - 1)
         return false;
     ++setIndex;
+    populateSetUrls();
     return true;
 }
 
+bool PictureToLoad::nextUrl()
+{
+    /* If we are past the list of urls currently populated
+     * for this set, try to move to the next set.  This
+     * will repopulate the list of urls.
+     */
+    if (urlIndex == currentSetUrls.size() - 1) {
```

Comment:
A note, if tomorrow we are going to replace the 2 hardcoded urlTemplates with a list.
What happens if i have no url templates defined?
`currentSetUrls.size() - 1` will be -1, `urlIndex` will be 0 and this will infinite loop up to MAX_INT.

Reply:
Agreed.  This was done at first by replicating the methods for iterating through the sets, where this same check is done.  I can take a look at that one as well (for example if a card has 0 sets? not sure if that's possible.